### PR TITLE
Code cleanup and performance improvements - Part 2

### PR DESCRIPTION
**Pull request by brainplot**
I found a few more issues after a more thorough inspection of Clazy, that was disabled by default.

## Description
* **Avoid allocating temporary containers**
Calling `toLower()` on a `QString` will cause a new string to be allocated. Since the function is essentially invoked to perform a case-insensitive comparison, we can invoke the `contains()` method with the `Qt::CaseInsensitive` enum value to do just that, as opposed to calling `toLower()` on two different strings.

* **Remove unused variables**
For some reason there were completely unused variables in the code. One of them was even initialized with the result of a method, but never used nonetheless (maybe it was the source of a bug).

* **Use `QVariant::toUrl()` instead of `QVariant::value<QUrl>()`**
The analyzer was suggesting me to do this. I guess the latter is more performant since it's not a template :thinking:

* **Fix typo in parameter name**
This is a very minor thing. There was a typo in the declaration of a function (`erroMessage` instead of `errorMessage`), which was spelt correctly in the definition of the same function, and the analyzer reported the inconsistency between the two.

* **Improve type-safety**
Some zeros used to initialize a pointer were left out of my previous commit. I think they were introduced with the PR which adds more expiration presets.

## How has this been tested?
Compiled with all tests enabled (GUI tests included) and `-DWITH_DEV_BUILD` on and all tests passed.

## Types of changes
- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by phoerious** [comment]
Please revert the for loop changes. Qt containers will not detach as long as you don't call non-const functions on them or the reference counter is 1.

**Comment by brainplot** [comment]
@phoerious I just read about this feature which I didn't know about. However, isn't it better to explicitly state in code the intent of not copying a container instead of relying on an implementation detail?
This is just my thought on this. If it's wrong, I can revert my commit, no problem.

**Comment by phoerious** [comment]
We use asConst() whenever we are iterating over a non-const container and there is no reason for copying rvalues.
Implicit sharing is not an implementation detail but an integral part of the Qt API.

File: src/core/EntryAttachments.cpp [hunk 1]
```diff
@@ -37,7 +37,11 @@ bool EntryAttachments::hasKey(const QString& key) const
 
 QSet<QByteArray> EntryAttachments::values() const
```
Comment (droidmonkey) [inline src/core/EntryAttachments.cpp]: I actually question the validity of this function in general. It gives up the purpose of the EntryAttachments class. In fact, its only used in one place [Entry.cpp:L579](https://github.com/keepassxreboot/keepassxc/blob/82bf00de7389db3e52baf11c9295300ad943c643/src/core/Entry.cpp#L579) in a rather useless fashion!

**Comment by brainplot** [comment]
Ok, I'll revert the changes to the for-loops in my commit.
Do I have to revert my change to the `toLower()` call? Using `Qt:CaseInsensitive` will avoid the allocation of the new string and seems a more natural approach to the problem to me.

File: src/sshagent/SSHAgent.cpp [hunk 2]
```diff
@@ -38,7 +38,7 @@ SSHAgent::SSHAgent(QObject* parent)
 
 SSHAgent::~SSHAgent()
 {
-    for (const QSet<OpenSSHKey>& keys : m_keys.values()) {
+    for (const QSet<OpenSSHKey>& keys : m_keys) {
```
Comment (droidmonkey) [inline src/sshagent/SSHAgent.cpp]: Might as well change this to **const auto&**

**Comment by droidmonkey** [comment]
Keep the Qt:CaseInsensitive, that is a good change and makes the code more readable in general.

**Comment by phoerious** [comment]
I haven't reviewed the full code yet. But reverting the iterator changes is the most important point. You may use STL-style iterators for maps for efficiency, but it's extremely ugly. After all, we don't really have performance problems.

**Comment by brainplot** [comment]
@phoerious @droidmonkey I reverted my changes.

**Comment by brainplot** [comment]
I ran `make format` and noticed that the beautifier applied a lot of fixes to the code formatting.
I've gone through each file one by one with my difftool, and wrapped with `// clang-format` off and `// clang-format on` those lines that looked more readable with manual formatting rather than automatic formatting.

I did this because, when I made my first PR, I wanted to make sure that my code followed the project's coding style; when I ran `make format`, however, I ended up with lots of modified files in my working directory. Now people should be able to format their code automatically.

**Comment by droidmonkey** [comment]
Can I give like 5 hearts?

**Comment by droidmonkey** [comment]
Can you add these changes to your list:

```
Function
    isNull() const : bool
Found usages  (7 usages found)
    DatabaseOpenWidget.cpp  (1 usage found)
        169 if (!pw.isNull()) {
    DatabaseTabWidget.cpp  (2 usages found)
        160 if (!i.value().dbWidget->dbHasKey() && !(pw.isNull() && keyFile.isEmpty())) {
        199 if (!pw.isNull() || !keyFile.isEmpty()) {
    KdbxXmlReader.cpp  (1 usage found)
        734 if (!ref.first.isNull() && !ref.second.isNull()) {
    TestKeePass1Reader.cpp  (1 usage found)
        277 if (!password.isNull()) {
    TestKeys.cpp  (2 usages found)
        229 if (!error.isNull()) {
        287 if (!error.isNull()) {
```

QString::isNull() is inappropriate usage, should be QString::isEmpty()

**Comment by brainplot** [comment]
@droidmonkey Done! I also replaced a few checks of the form of `if (obj.size() > 0)` to `if (!obj.empty())`.
It looks more readable in my opinion.

**Comment by droidmonkey** [comment]
Please rebase onto the new develop. @phoerious please update your review.

**Comment by brainplot** [comment]
@droidmonkey done!

**Comment by droidmonkey** [comment]
Going to merge in the advanced search PR, then rebase this one and merge next.

**Comment by droidmonkey** [comment]
@brainplot major refactors are complete, please rebase and fixup where necessary.

**Comment by brainplot** [comment]
Alright, this one was a little tricky to rebase because there was a good bunch of conflicts to deal with; but I managed it to do it correctly, I believe, since I ran all the tests and they all passed.

I made one more change at the end: there was a variable declared in the `.cpp` file without being enclosed in any scope. Moreover, it was only used in a single function. I moved it into the function it was used it and made it `static` so that it gets initialized only once and not every time the function is invoked.
```diff
-// Escape common regex symbols except for *, ?, and |
-auto regexEscape = QRegularExpression(R"re(([-[\]{}()+.,\\\/^$#]))re");
-
 QRegularExpression convertToRegex(const QString& string, bool useWildcards, bool exactMatch, bool caseSensitive)
 {
     QString pattern = string;

+    // Escape common regex symbols except for *, ?, and |
+    static auto regexEscape = QRegularExpression(R"re(([-[\]{}()+.,\\\/^$#]))re");
+
     // Wildcard support (*, ?, |)
     if (useWildcards) {
         pattern.replace(regexEscape, "\\\\1");
         pattern.replace("*", ".*");
         pattern.replace("?", ".");
     }

     // Exact modifier
     if (exactMatch) {
         pattern = "^" + pattern + "$";
     }

     auto regex = QRegularExpression(pattern);
     if (!caseSensitive) {
         regex.setPatternOptions(QRegularExpression::CaseInsensitiveOption);
     }

     return regex;
 }
```

**Comment by brainplot** [comment]
Alright, this one was a little tricky to rebase because there was a good bunch of conflicts to deal with; but I managed it to do it correctly, I believe, since I ran all the tests and they all passed.

I made one more change at the end: there were two variables that could have their scope reduced by making them have static storage.
In one case, the variable was declared in the `.cpp` file as file-scoped but it was only used in a single function. I moved the variable into the function it was used in and made it `static` so that it'll get initialized only once regardless of how many times the function is called.
```diff
-// Escape common regex symbols except for *, ?, and |
-auto regexEscape = QRegularExpression(R"re(([-[\]{}()+.,\\\/^$#]))re");
-
 QRegularExpression convertToRegex(const QString& string, bool useWildcards, bool exactMatch, bool caseSensitive)
 {
     QString pattern = string;

+    // Escape common regex symbols except for *, ?, and |
+    static auto regexEscape = QRegularExpression(R"re(([-[\]{}()+.,\\\/^$#]))re");
+
     // Wildcard support (*, ?, |)
     if (useWildcards) {
         pattern.replace(regexEscape, "\\\\1");
         pattern.replace("*", ".*");
         pattern.replace("?", ".");
     }

     // Exact modifier
     if (exactMatch) {
         pattern = "^" + pattern + "$";
     }

     auto regex = QRegularExpression(pattern);
     if (!caseSensitive) {
         regex.setPatternOptions(QRegularExpression::CaseInsensitiveOption);
     }

     return regex;
 }
```

In the other case, the variable was declared as file-scoped and used in two different functions so I couldn't move it anywhere. However, making it `static` will tell the compiler that the variable is only used in the current file, which will prevent it to "leak" into other translation units.
```diff
-MainWindow* g_MainWindow = nullptr;
+static MainWindow* g_MainWindow = nullptr;
 MainWindow* getMainWindow()
 {
     return g_MainWindow;
 }
 
 MainWindow::MainWindow()
     : m_ui(new Ui::MainWindow())
     , m_trayIcon(nullptr)
     , m_appExitCalled(false)
     , m_appExiting(false)
 {
     g_MainWindow = this;
 
     m_ui->setupUi(this);
 
    /* ... */
}
```

**Comment by droidmonkey** [comment]
I did another rebase. I greatly simplified the use of clang-format off/on around signal/slot connections. Instead of isolating problem ones, I just wrapped the whole block with them. I fixed other areas where the clang-format comment was causing readability to go down.

**Comment by brainplot** [comment]
I thought about doing that as well, but since turning clang-format off on a line basis will prevent it to fix formatting issues that may be present on the whole line, I just wanted to turn it off in the narrowest scope possible. I was even considering something like this:
```cpp
connect(selectionModel(),
	// clang-format off
	SIGNAL(selectionChanged(QItemSelection,QItemSelection)),
	// clang-format on
	SLOT(emitEntrySelectionChanged())
	);
```
but arguably that looks even uglier.

Regarding to your changes, how do I integrate them into mine so that I can then rebase my branch?

**Comment by droidmonkey** [comment]
Sorry I was waiting for my local tests to finish building, its pushed now (to your branch).

**Comment by brainplot** [comment]
Yeah, that sounds reasonable.

Can I rebase now or is there any work that needs to be done first?

**Pull request merged**